### PR TITLE
[SPARK-13616][SQL] Let SQLBuilder convert logical plan without a project on top of it

### PR DESCRIPTION
**Pull request by viirya**
JIRA: https://issues.apache.org/jira/browse/SPARK-13616
## What changes were proposed in this pull request?

It is possibly that a logical plan has been removed `Project` from the top of it. Or the plan doesn't has a top `Project` from the beginning because it is not necessary. Currently the `SQLBuilder` can't convert such plans back to SQL. This change is to add this feature.
## How was this patch tested?

A test is added to `LogicalPlanToSQLSuite`.

### CODE REVIEW

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala [hunk 1]
```diff
@@ -78,6 +78,27 @@ class SQLBuilder(logicalPlan: LogicalPlan, sqlContext: SQLContext) extends Loggi
...
+    if (topNode) {
+      node match {
+        case d: Distinct => toSQL(node)
+        case p: Project => toSQL(node)
+        case a: Aggregate => toSQL(node)
+        case s: Sort => toSQL(node)
+        case r: RepartitionByExpression => toSQL(node)
+        case _ =>
+          build(
+          "SELECT",
```
Comment (rxin) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala]: indent

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/LogicalPlanToSQLSuite.scala [hunk 2]
```diff
@@ -157,6 +186,16 @@ class LogicalPlanToSQLSuite extends SQLBuilderTest with SQLTestUtils {
       "SELECT x.key, COUNT(*) FROM parquet_t1 x JOIN parquet_t1 y ON x.key = y.key group by x.key")
   }
 
+  test("join plan") {
+    val originalSql = "SELECT x.key FROM parquet_t1 x JOIN parquet_t1 y ON x.key = y.key"
```
Comment (rxin) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/LogicalPlanToSQLSuite.scala]: maybe you should construct a plan using dataframe?

**Comment by rxin** [comment]
Thanks - going to merge this in master.

cc @liancheng

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala [hunk 3]
```diff
@@ -78,6 +78,27 @@ class SQLBuilder(logicalPlan: LogicalPlan, sqlContext: SQLContext) extends Loggi
     }
   }
 
+  private def toSQL(node: LogicalPlan, topNode: Boolean): String = {
```
Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala]: I think the only missing cases are `Join`. All the other cases have been supported. It breaks multiple cases. I plan to revert this change back. cc @liancheng

Comment (viirya) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala]: Yea, looks like so. We only need to take care of Join for this purpose.

Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala]: Actually, I am also afraid that the test case here for `Join` is not valid. I guess `Join` is the top layer node only when users are using DataFrame/DataSet APIs. If users are using SQL, `Join` will not be the top layer node, right?

If we want to convert the analyzed logical plan generated from DataFrame API to SQL, this is not the only issue. I also hit a couple of issues before. @liancheng , do you think we should support them?

Comment (viirya) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala]: Instead of reverting this, I think we can simply change the pattern cases to let it match Join to do the.

Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala]: The analyzed plan of this query: 
```
"SELECT x.key FROM parquet_t1 x JOIN parquet_t1 y ON x.key = y.key"
``` 
```
== Analyzed Logical Plan ==
key: bigint
Project [key#18L]
+- Join Inner, Some((key#18L = key#23L))
   :- SubqueryAlias x
   :  +- SubqueryAlias parquet_t1
   :     +- Relation[key#18L,value#19] ParquetRelation
   +- SubqueryAlias y
      +- SubqueryAlias parquet_t1
         +- Relation[key#23L,value#24] ParquetRelation
```

Comment (viirya) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala]: Yea, I see. Join is the top node only when using DataFrame API. But the original purpose of this change is not intended to support converting logical plan of DataFrame to SQL. The test case is just used to construct the Join plan needed to test it. This change is introduced to support the option to convert a logical plan not beginning from the top of the whole plan.

Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala]: Actually, I am unable to find any analyzed plan (generated by the SQL parser) that breaks the existing solution. The existing solution already covers all the cases, I think.

Comment (viirya) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala]: plain Join?

Comment (viirya) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala]: I can't see why we want to set a constraint that we can only convert a whole logical plan generated from SQL query, not a partial and legal logical plan from it.

Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala]: If my understanding is not wrong, you want to support SQL generation from a partial analyzed plan. If so, do you have any use case?

So far, our SQL generation assumes the plan is complete and generated by the SQL parser. If not, I am afraid, it could generate a wrong SQL statement.

Comment (viirya) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala]: If the algorithm can convert an analyzed logical plan back to SQL. I can't see why we can't let it convert its child plan back to SQL.

Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala]: After rethinking this issue, I believe SQL generation of partial **optimized** plans is very important for federation to RDBMS. However, SQL generation of optimized plans is much more complex. So far, it is not the purpose. I am wondering if SQL generation of partial **analyzed** plans is needed?

Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala]: I have to go to bed now. Either is fine to me. : )

Comment (viirya) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala]: At least for the purpose of SQLBuilder, I think the capability to convert a child of an analyzed logical plan is important. You can convert the children plans of the original plan back to several SQLs and perform these SQLs on different RDBMS, for example.

Comment (viirya) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala]: ok. Maybe we can discuss more offline.

Comment (liancheng) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala]: @viirya @gatorsmile Thanks for your work and discussions!  The initial motivation for implementing SQL generation is for better native view support. This makes the following constraints reasonable for the initial version of SQL generation in Spark 2.0:

1. The target logical plan must be parsed from a valid HiveQL query statement

   So that the structure of the original SQL statement is preserved as much as possible (for example, subquery scoping information). This makes mapping from plan fragments to their SQL representations much easier.

1. The target logical plan must be _fully_ resolved

   Basically you can't guarantee that an unresolved / partially resolved logical plan is actually valid. And they may contain unwanted auxiliary expressions / operators like `UnresolvedAlias`, which further complicate SQL generation.

1. The target logical plan should NOT be optimized

   Similar to 1.

Also, for native view support, generating optimal SQL statements is NOT a requirement.  It's OK that we generate verbose and inefficient SQL statements as long as the Catalyst optimizer can optimize them at runtime.

Ideally, I'd like to remove constraints 1 and 3 in the future, so that SQL generation can be applied to wider scenarios, e.g., random query testing.  But for Spark 2.0, let's focus on fully resolved, non-optimized logical plans parsed from valid HiveQL first.  Correctness and test coverage is more important at the current stage.

So my suggestion is:

1. Revert this change
1. Revisit it after we finish SQL generation support of all major language structures

   Window functions and generators are not supported yet.  I'm afraid we may miss important cases if we try to do this work now.

Comment (liancheng) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala]: BTW, more details and background about SQL generation and native view support can be found in [SPARK-11012][1].

[1]: https://issues.apache.org/jira/browse/SPARK-11012

Comment (liancheng) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala]: @gatorsmile You mentioned that you found some regressions. Did you mean some test cases in the current master branch are failing, or you constructed new test cases that failed this PR?

Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala]: I saw the test cases in HiveCompatibilitySuite failed. 
As @dilipbiswal said, `Limit` is one of the issues. `Filter` is another one. Besides these two, I also saw more. Thus, I did not continue to fix all of them.

Comment (liancheng) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala]: I see. So this PR decreases SQL generation coverage but didn't cause actual build failures, right? Still I'm going to revert it by merging #11539.

Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala]: Yeah
